main/sig_finder/sig_finder.cc

```diff
@@ -8,110 +9,133 @@ using namespace std;
-            // - Hit the start of the file.
-            line_end = prev_line_end;
-            prev_line_end = source.rfind('\n', line_end - 1);
-            auto multiLineSigEndPos = line.data() - source.data() + multiLineSigEndSize;
-            while (prev_line_end != source.npos) {
-                auto line_start = prev_line_end + 1;
-                line = source.substr(line_start, line_end - line_start);
-                line = absl::StripAsciiWhitespace(line);
-
-                // Found the start of the multi-line sig we were looking for.
-                if (absl::StartsWith(line, multiLineSigStart)) {
-                    uint32_t offset = line.data() - source.data();
-                    sigLocOffsets = core::LocOffsets{offset, offset + singleLineSigSize};
-                    bodyLocOffsets =
-                        core::LocOffsets{offset + singleLineSigSize + 1, static_cast<uint32_t>(multiLineSigEndPos)};
-                    break;
-                }
-
-                if (absl::StartsWith(line, alreadyFinalSig)) {
-                    break;
-                }
-
-                // We hit an `end` in a place we didn't expect.  Bail.
-                if (absl::StartsWith(line, multiLineSigEnd)) {
-                    break;
-                }
-
-                line_end = prev_line_end;
-                prev_line_end = source.rfind('\n', line_end - 1);
-            }
-
-            // However we exited the above loop, we are done.
-            break;
+void SigFinder::preTransformMethodDef(core::Context ctx, ast::ExpressionPtr &tree) {
+    if (this->result_.has_value()) {
+        if (this->result_->origSend->loc.endPos() <= tree.loc().beginPos() &&
+            tree.loc().endPos() <= queryLoc.beginPos()) {
+            // There is a method definition between the current result sig and the queryLoc,
+            // so the sig we found is not for the right method.
+            this->result_ = nullopt;
         }
+    }
+}
 
-        // We found a comment between a sig and a method def, look one line higher
-        if (absl::StartsWith(line, commentLineStart)) {
-            line_end = prev_line_end;
-            prev_line_end = source.rfind('\n', line_end - 1);
-            continue;
+void SigFinder::postTransformRuntimeMethodDefinition(core::Context ctx, ast::ExpressionPtr &tree) {
+    if (this->result_.has_value()) {
+        if (this->result_->origSend->loc.endPos() <= tree.loc().beginPos() &&
+            tree.loc().endPos() <= queryLoc.beginPos()) {
+            // There is a method definition between the current result sig and the queryLoc,
+            // so the sig we found is not for the right method.
+            this->result_ = nullopt;
         }
+    }
+}
+
+void SigFinder::preTransformSend(core::Context ctx, ast::ExpressionPtr &tree) {
+    auto &send = ast::cast_tree_nonnull<ast::Send>(tree);
+
+    if (!resolver::TypeSyntax::isSig(ctx, send)) {
+        return;
+    }
```

Comment:
A general comment (not necessarily attached to this line) -- I have very little intuition for whether using an AST walk to find `sig`s is slower/faster/about the same as compared to the current string matching method. I trust that this has been tested for perf, though.

Reply:
Some notes:

- This code is only running in places where speed isn't absolutely critical (error messages, LSP requests)
- This code is more correct (source string parsing is inherently error prone)

I have not tested it for performance, again because our normal mechanism for testing this (just run sorbet a bunch) wouldn't exercise this codepath unless we introduced tons of final method errors.

Reply:
But my intuition is that ASTs are faster, as you can walk the tree without ever having to allocate or do a linear string search.